### ISSUE TITLE
Use Blackfire.io profiler for measurements and analyze WordPress behavior

### ISSUE THREAD

**Issue by lolautruche** [WordPress/performance#43]
[Blackfire.io](https://blackfire.io/docs/introduction) is willing to help the performance initiative by providing:

- A free Enterprise subscription, including all add-ons and free training (by myself);
- Developers resources: myself and @gilzow.

Blackfire provides [Profiling](https://blackfire.io/docs/introduction#profiling), [Monitoring](https://blackfire.io/docs/introduction#monitoring) and [automated Performance Testing](https://blackfire.io/docs/introduction#testing).

In the longer run, WordPress.org would have its own organization.

We can use Blackfire to help reach most of the performance targets (as long as PHP is involved). 
Blackfire can [help us understand how WP is behaving](https://blackfire.io/docs/profiling-cookbooks/understanding-call-graphs), [what happens through time using the timeline](https://blackfire.io/docs/profiling-cookbooks/understanding-timelines), and give us hints on how to improve its behavior. We can also [compare profiles, to precisely measure the impact of the changes](https://blackfire.io/docs/profiling-cookbooks/understanding-comparisons).

With automated performance tests, we may write up [metrics](https://blackfire.io/docs/testing-cookbooks/metrics), [tests](https://blackfire.io/docs/testing-cookbooks/tests), and [scenarios](https://blackfire.io/docs/builds-cookbooks/scenarios) to [run on a periodic basis](https://blackfire.io/docs/builds-cookbooks/builds-periodic), or even on [each PR/push](https://blackfire.io/docs/integrations/git/github#documentation).

Let's discuss this :slightly_smiling_face:

**Comment by adamsilverstein** [comment]
Hi @lolautruche 👋🏼  thanks for opening this issue. Exciting to hear you will be contributing to the WordPress core performance effort!

I'm not too familiar with your service/product, so thanks for all the links - that helped. 

Can you to describe how you would see Backfire.io fitting into our effort? Would we use it to test the impact of proposed changes when developing features?  

We've also discussed bring some sort of monitoring into core that you might be able to contribute to directly if you are interested.

If you are able, please attend the performance group meeting when you can to introduce yourself and Blackfire.io - Tuesdays in the WordPress core #performance channel (signup https://make.wordpress.org/chat/, exact time - https://make.wordpress.org/meetings/)

**Comment by westonruter** [comment]
We've used Blackfire on the AMP plugin to identify some initial performance bottlenecks, but we've stalled on adding it to CI: https://github.com/ampproject/amp-wp/issues/3361. /cc @schlessera

**Comment by lolautruche** [comment]
@adamsilverstein We may use the Profiler to analyze all the requests, regardless what we're aiming to optimize. In fact, as long as PHP is involved in the operation we want to improve, Blackfire can help as it gives facts and figures on how WP is doing **exactly**.
We'll see each and every function calls, how many times they've been called, which functions called whichever function, etc. Furthermore, we'll be able to see how much I/O time/CPU time/Memory each function took, what were the SQL requests involved, etc.
Once a patch is ready, we can profile again and compare the new profile with new previous one, to measure the impact.

Given that any performance target, any change would be covered by profiles, we would obtain substantial results, because we'd base our work on facts and figures 🙂 .

As for the core monitoring you're mentioning, what is it about exactly? Would it be real e.g., defining key scenarios with performance tests, and comparison for each feature/fix? 
If so, this is what we call Synthetic User Monitoring, which Blackfire can cover, and that I'd gladly contribute to!

Regarding the performance group, I'm actually already following the Slack channel and was present at the meeting last week 😉 .
I won't be able to attend tomorrow, but I guess @gilzow will be.

**Comment by lolautruche** [comment]
> We've used Blackfire on the AMP plugin to identify some initial performance bottlenecks, but we've stalled on adding it to CI: https://github.com/ampproject/amp-wp/issues/3361. /cc @schlessera

I'd gladly help you on this!
I noticed that the issue is 2y old, I hope I'm not getting there after the battle 😊 .

**Comment by schlessera** [comment]
I've been in contact with Blackfire for a while now to get traction in WordPress Core. There is a `WordPress` enterprise environment that was kindly provided for open source WordPress Core work. I won't share the credentials here, but @lolautruche I'm sure you have internal access to find that.

I was in contact with multiple Core committers over time letting them know about this possibility, but enthusiasm for profiling has been ... low.

Hopefully, with the new `performance` focus and team, this can be picked up again.

**Comment by lolautruche** [comment]
Thanks @schlessera for reminding me of that environment. I have indeed all the details here.
I'll be able to add new users to this environment. We should organize this somehow, as I'd need email addresses.

**Comment by eugene-manuilov** [comment]
@lolautruche is it possible to see profiling results on a local machine without going to the blackfire site? The ability to observe profiling results while working on a feature on a local machine is one of the key factors of success for this project. Without it, the profiling will worth nothing if 99% of developers won't be able to see results just because they don't have access to the Blackfire account.

**Comment by lolautruche** [comment]
@eugene-manuilov No, you have to access the Blackfire website to view the profile. I understand the issue, but we may add as many developers as needed, as long as they use it :wink:

**Comment by eugene-manuilov** [comment]
Thanks, @lolautruche, but I think it will be problematic and easily become a bottleneck.

**Comment by schlessera** [comment]
@eugene-manuilov - A profiling run can be shared as a publically accessible URL that anyone can go through even without an account.

Example link (just click on it to try for yourself): https://blackfire.io/profiles/compare/bb84da75-a28b-4c66-ab88-a352c8c56525/graph

**Comment by eugene-manuilov** [comment]
@schlessera this is good, but it still doesn't solve the local development problem.

**Comment by josephscott** [comment]
I think there is some confusion here.  Using Blackfire on a consistent setup tracking the master branch can help us identify commits that might be hurting performance.  At the same time, devs can work against the same type of setup locally ( or on a separate server some where else ).  The only difference being that they won't have access to the Blackfire tools for their local setup.

**Comment by felixarntz** [comment]
For reference, we now have #64 which is more holistically about which performance tooling to use for our initial project(s). Potentially you'd want to chime in there as well.

**Comment by swissspidy** [comment]
I don't think there's anything actionable left to be done in this ticket. Blackfire has been regularly used by contributors and usage in the WordPress context is documented in the team's handbook as well.
